recipes-extended/xen/files/libxl-blktap3-support.patch

```diff
@@ -0,0 +1,358 @@
++    }
++
++    *disk++ = '\0';
++
++    err = blktap_find(type, disk, &tap);
++    if (err < 0) {
++        /* returns -errno */
++        LOGEV(ERROR, -err, "Unable to find type %s disk %s", type, disk);
++        return ERROR_FAIL;
++    }
++
++    err = tap_ctl_destroy(tap.pid, tap.minor, 0, NULL);
++    if (err < 0) {
++        LOGEV(ERROR, -err, "Failed to destroy tap device id %d minor %d",
++              tap.pid, tap.minor);
++        return ERROR_FAIL;
++    }
++
++    return 0;
++}
++
++/*
++ * Local variables:
++ * mode: C
++ * c-basic-offset: 4
++ * indent-tabs-mode: nil
++ * End:
++ */
+Index: xen-4.9.1/tools/libxl/Makefile
+===================================================================
+--- xen-4.9.1.orig/tools/libxl/Makefile
++++ xen-4.9.1/tools/libxl/Makefile
+@@ -20,17 +20,17 @@ LIBUUID_LIBS += -luuid
+ endif
+ 
+ LIBXL_LIBS =
+-LIBXL_LIBS = $(LDLIBS_libxentoollog) $(LDLIBS_libxenevtchn) $(LDLIBS_libxenctrl) $(LDLIBS_libxenguest) $(LDLIBS_libxenstore) $(LDLIBS_libblktapctl) $(PTYFUNCS_LIBS) $(LIBUUID_LIBS)
++LIBXL_LIBS = $(LDLIBS_libxentoollog) $(LDLIBS_libxenevtchn) $(LDLIBS_libxenctrl) $(LDLIBS_libxenguest) $(LDLIBS_libxenstore) $(PTYFUNCS_LIBS) $(LIBUUID_LIBS)
+ ifeq ($(CONFIG_LIBNL),y)
+ LIBXL_LIBS += $(LIBNL3_LIBS)
+ endif
+ 
++CFLAGS_LIBXL += -I/usr/include/blktap
+ CFLAGS_LIBXL += $(CFLAGS_libxentoollog)
+ CFLAGS_LIBXL += $(CFLAGS_libxenevtchn)
+ CFLAGS_LIBXL += $(CFLAGS_libxenctrl)
+ CFLAGS_LIBXL += $(CFLAGS_libxenguest)
+ CFLAGS_LIBXL += $(CFLAGS_libxenstore)
+-CFLAGS_LIBXL += $(CFLAGS_libblktapctl) 
+ ifeq ($(CONFIG_LIBNL),y)
+ CFLAGS_LIBXL += $(LIBNL3_CFLAGS)
+ endif
+@@ -47,9 +47,9 @@ LIBXLU_LIBS = $(LDLIBS_libxenlight)
+ 
+ LIBXL_OBJS-y = osdeps.o libxl_paths.o libxl_bootloader.o flexarray.o
+ ifeq ($(LIBXL_BLKTAP),y)
+-LIBXL_OBJS-y += libxl_blktap2.o
++LIBXL_OBJS-y += libxl_blktap3.o
+ else
+-LIBXL_OBJS-y += libxl_noblktap2.o
++LIBXL_OBJS-y += libxl_noblktap3.o
+ endif
+ 
+ ifeq ($(CONFIG_LIBNL),y)
+Index: xen-4.9.1/tools/libxl/libxl_types_internal.idl
+===================================================================
+--- xen-4.9.1.orig/tools/libxl/libxl_types_internal.idl
++++ xen-4.9.1/tools/libxl/libxl_types_internal.idl
+@@ -27,6 +27,7 @@ libxl__device_kind = Enumeration("device
+     (10, "QUSB"),
+     (11, "9PFS"),
+     (12, "VWIF"),
++    (13, "VBD3"),
+     ])
+ 
+ libxl__console_backend = Enumeration("console_backend", [
+Index: xen-4.9.1/tools/libxl/libxl_device.c
+===================================================================
+--- xen-4.9.1.orig/tools/libxl/libxl_device.c
++++ xen-4.9.1/tools/libxl/libxl_device.c
+@@ -297,11 +297,6 @@ static int disk_try_backend(disk_try_bac
+ 
+         /* Note: keep going if the disk is a cdrom. tapdisk can handle them as "raw" */
+ 
+-        if (!libxl__blktap_enabled(a->gc)) {
+-            LOG(DEBUG, "Disk vdev=%s, backend tap unsuitable because blktap "
+-                       "not available", a->disk->vdev);
+-            return 0;
+-        }
```

Comment:
This check is not required now as `libxl__blktap_enabled` always returns '1', because in blktap3 blkfront directly contacts tapdisk (unlike blktap2 where the path is blkfront --> blkback -->blktap2-kernel-module --> tapdisk). In blktap2 this check made sense as blktap2-kernel-module was involved in the path. However, this check is required by tapdisk/tap-ctl processes as blktap2-kernel-module is still used here and `tap_ctl_check_blktap` is being done in this place as required.

P.S: This is in accordance with the base patch for blktap3, given here : https://github.com/qiaohaiming/blktap3/blob/master/libxl.diff

Reply:
Although the above lines are redundant, I will revert back the changes for the sake of blktap2 retro-compatibility support.
The implementation searching for "blktap/control" instead of "blktap-control" is already fixed, I think you might have looked at older code.
Thanks.